File: packages/secure-storage/src/SecureStorageModule.ts
import type { RynBridge } from '@rynbridge/core';
import type {
  SecureStorageGetResponse,
  SecureStorageHasResponse,
} from './types.js';

const MODULE = 'secure-storage';

export class SecureStorageModule {
  private readonly bridge: RynBridge;

  constructor(bridge: RynBridge) {
    this.bridge = bridge;
  }

  async get(key: string): Promise<string | null> {
    const result = await this.bridge.call(MODULE, 'get', { key });
    return (result as unknown as SecureStorageGetResponse).value;
  }

  async set(key: string, value: string): Promise<void> {
    await this.bridge.call(MODULE, 'set', { key, value });
  }

  async remove(key: string): Promise<void> {
    await this.bridge.call(MODULE, 'remove', { key });
  }

  async has(key: string): Promise<boolean> {
    const result = await this.bridge.call(MODULE, 'has', { key });
    return (result as unknown as SecureStorageHasResponse).exists;
  }
}

File: android/secure-storage/src/main/kotlin/io/rynbridge/securestorage/SecureStorageModule.kt
package io.rynbridge.securestorage

import android.content.Context
import io.rynbridge.core.*

class SecureStorageModule(provider: SecureStorageProvider) : BridgeModule {
    constructor(context: Context) : this(DefaultSecureStorageProvider(context))

    override val name = "secure-storage"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "get" to { payload ->
            val key = payload["key"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: key")
            val value = provider.get(key)
            if (value != null) {
                mapOf("value" to BridgeValue.string(value))
            } else {
                mapOf("value" to BridgeValue.nullValue())
            }
        },
        "set" to { payload ->
            val key = payload["key"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: key")
            val value = payload["value"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: value")
            provider.set(key, value)
            emptyMap()
        },
        "remove" to { payload ->
            val key = payload["key"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: key")
            provider.remove(key)
            emptyMap()
        },
        "has" to { payload ->
            val key = payload["key"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: key")
            val exists = provider.has(key)
            mapOf("exists" to BridgeValue.bool(exists))
        }
    )
}
